feat(ui): start using Datatables to generate Objects table

tomrtk opened the pull request:
Closes #60, closes #80.

tomrtk committed:
refactor(ui): implemented use of Datatable for objects

Implemented use of datatable with pagination for objects table on job page.

tomrtk commented:
Current state of table:
![image](https://user-images.githubusercontent.com/43438127/124739492-32d69700-df1a-11eb-9002-1307edd84d43.png)

R0flcopt3r commented on src/core/api/api.py:
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
+    project_id: int = Path(..., ge=1),
+    job_id: int = Path(..., ge=1),
+    start: int = Query(0, ge=0),
+    lenght: int = Query(10, ge=1),
```
```suggestion
    length: int = Query(10, ge=1),
```

tomrtk replied:
This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

R0flcopt3r commented on src/core/api/api.py:
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
+    -------
+    Dict[str, Any]
+        Containing total number of object in job and part of objects from
+        `start` to `start + lenght`.
```
```suggestion
        `start` to `start + length`.
```

tomrtk replied:
This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

R0flcopt3r commented on src/core/api/api.py:
```diff
@@ -611,3 +611,43 @@ async def get_object_image(
+        If project or job id provided is not found. Status code: 422.
+    """
+    try:
+        data = services.get_job_objects(project_id, job_id, start, lenght)
```
```suggestion
        data = services.get_job_objects(project_id, job_id, start, length)
```

R0flcopt3r commented on src/core/services.py:
```diff
@@ -508,3 +508,55 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
+
+    response: Dict[str, Any] = {}
+    response["total_objects"] = len(job._objects)
+    response["data"] = job._objects[start : start + lenght]
```
```suggestion
    response["data"] = job._objects[start : start + length]
```

tomrtk replied:
This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

R0flcopt3r commented on src/core/services.py:
```diff
@@ -508,3 +508,55 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
+
+
+def get_job_objects(
+    project_id: int, job_id: int, start: int, lenght: int
```
```suggestion
    project_id: int, job_id: int, start: int, length: int
```

tomrtk replied:
This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

R0flcopt3r commented on src/ui/projects/api.py:
```diff
@@ -449,3 +449,37 @@ def check_api(self) -> bool:
         return True
+
+    def get_objects(
+        self, project_id: int, job_id: int, start: int, lenght: int
```
```suggestion
        self, project_id: int, job_id: int, start: int, length: int
```

tomrtk replied:
This PR is based upon #72 and this is from there. It is all ready fixed in that PR.

R0flcopt3r commented on src/ui/projects/utils.py:
```diff
@@ -0,0 +1,22 @@
+
+
+@validate_arguments
+def validate_int(
```
Wouldn't it be better if it returned an `Optional[int]`?

R0flcopt3r commented on src/ui/src/js/projects.js:
```diff
@@ -23,6 +23,8 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+require("datatables.net")(window, $);
```
does this grab scripts from external sites?

tomrtk replied:
@MindTooth can probably answer this better. datatables is installed with npm and this indicate it is needed, similar to import, but the import syntax did not work for me. So I stayed with what worked, not using time to get it working.

MindTooth replied:
This syntax is called CommonJS and is to be regardes as obsolete.  Now we use what’s called ESM.  `import` is newer and should be used.

Regarding `.net`, the package has it in its name. 😅

tomrtk commented:
Remaining work:
- [x] ~~Use stats from core when #75 is merged. This can maybe be it's own issue and PR.~~ new issue #80 
- [x] Get preview popup to work correctly.

Feel free to work on popup if anyone has an idea of a good way of doing it.

And also please test run branch.

tomrtk committed:
fix(ui): fix error in logging on api client

tomrtk committed:
refactor(ui): make a more generic error handling template

tomrtk committed:
fix: validate route parameter for object preview

MindTooth commented on src/ui/src/js/projects.js:
```diff
@@ -23,6 +23,8 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+var dt = require("datatables.net")(window, $);
```
```suggestion
import dt from "datatables.net";
dt(window, $);
```

So we use imports all over. 😉

Kad1n reviewed (commented):
Seems to be some cases in `projects.py` that is not covered by tests. They are mostly error checks. Should perhaps write a quick test to cover them.

Otherwise this PR looks good :+1:

R0flcopt3r commented on src/core/services.py:
```diff
@@ -563,7 +563,6 @@ def get_job_objects(
     response: Dict[str, Any] = {}
     response["total_objects"] = len(job._objects)
     response["data"] = job._objects[start : start + length]
-    print(response["data"])
```
This should probably be removed?

tomrtk replied:
??
```diff
-    print(response["data"])
```

R0flcopt3r replied:
brane go brrr.....

R0flcopt3r commented on src/ui/projects/api.py:
```diff
@@ -449,3 +447,39 @@ def check_api(self) -> bool:
+            dct = response.json()
+            objects = [Object(**d) for d in dct["data"]]
+            total_object_count = dct["total_objects"]
+            return objects, total_object_count
```
Total amount of object should also be in `Job`/`JobBare` as part of `JobStat` if I remember correctly.

tomrtk replied:
The total amount is needed to be part of response to Datatable, so the object endpoint is returning it also.

R0flcopt3r replied:
I see!

tomrtk committed:
fix(ui): bug in parsing of video paths from user

tomrtk committed:
test(ui): improved test coverage of ui

tomrtk commented:
Added some more tests. Main contribution of PR should almost have 100% coverage. Some of the fixes and improvement in other parts of `ui` code base are partial testes since they do not have any tests from before.

tomrtk committed:
refactor: start use `job.stats` in ui and stop sending `job._objects'

- `core` api stop sending list of objects as part of a `Job`.
- start using the `job.stats` in `ui` model and template.
Closing #80

MindTooth commented on src/ui/projects/api.py:
```diff
@@ -449,3 +447,39 @@ def check_api(self) -> bool:
+    def get_objects(
+        self, project_id: int, job_id: int, start: int, length: int
+    ) -> Optional[Tuple[List[Object], int]]:
+        """Get a list of Objects for `job_id` from `start` to `start + lenght`.
```
```suggestion
        """Get a list of Objects for `job_id` from `start` to `start + length`.
```

MindTooth commented on src/ui/src/js/projects.js:
```diff
@@ -23,6 +23,9 @@ window.$ = window.jQuery = jQuery;
 
 import "jstree";
 
+import dt from "datatables.net";
```
```suggestion
import 'datatables.net'
import 'datatables.net-dt'
```

Seemed they were both needed.  I'll test tomorrow.

MindTooth commented on tests/unit/ui/test_utils.py:
```diff
@@ -0,0 +1,20 @@
+from ui.projects.utils import validate_int, SQLITE_INTEGER_LIMIT
+
+
+def test_vaidate_int() -> None:
```
```suggestion
def test_validate_int() -> None:
```

tomrtk committed:
fix: typo in code and docstring

Co-authored-by: Birger J. Nordølum <contact@mindtooth.no>
